core/src/main/java/org/apache/gravitino/listener/ModelEventDispatcher.java

```diff
@@ -0,0 +1,400 @@
+import org.apache.gravitino.listener.api.event.DeleteModelEvent;
+import org.apache.gravitino.listener.api.event.DeleteModelVersionEvent;
+import org.apache.gravitino.listener.api.event.GetModelEvent;
+import org.apache.gravitino.listener.api.event.GetModelVersionEvent;
+import org.apache.gravitino.listener.api.event.LinkModelVersionEvent;
+import org.apache.gravitino.listener.api.event.ListModelEvent;
+import org.apache.gravitino.listener.api.event.ListModelVersionsEvent;
+import org.apache.gravitino.listener.api.event.RegisterModelEvent;
+import org.apache.gravitino.listener.api.info.ModelInfo;
+import org.apache.gravitino.listener.api.info.ModelVersionInfo;
+import org.apache.gravitino.model.Model;
+import org.apache.gravitino.model.ModelCatalog;
+import org.apache.gravitino.model.ModelVersion;
+import org.apache.gravitino.utils.PrincipalUtils;
+
+/**
+ * {@code ModelEventDispatcher} is a decorator for {@link ModelDispatcher} that not only delegates
+ * model operations to the underlying catalog dispatcher but also dispatches corresponding events to
+ * an {@link org.apache.gravitino.listener.EventBus} after each operation is completed. This allows
+ * for event-driven workflows or monitoring of model operations.
+ */
+public class ModelEventDispatcher implements ModelDispatcher {
+  private final EventBus eventBus;
+  private final ModelDispatcher dispatcher;
+
+  /**
+   * Constructs a {@link ModelEventDispatcher} with a specified EventBus and {@link
+   * ModelDispatcher}.
+   *
+   * @param eventBus The EventBus to which events will be dispatched.
+   * @param dispatcher The underlying {@link ModelDispatcher} that will perform the actual model
+   *     operations.
+   */
+  public ModelEventDispatcher(EventBus eventBus, ModelDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  /**
+   * Register a model in the catalog if the model is not existed, otherwise the {@link
+   * ModelAlreadyExistsException} will be thrown. The {@link Model} object will be created when the
+   * model is registered, users can call {@link ModelCatalog#linkModelVersion(NameIdentifier,
+   * String, String[], String, Map)} to link the model version to the registered {@link Model}.
+   *
+   * @param ident The name identifier of the model.
+   * @param comment The comment of the model. The comment is optional and can be null.
+   * @param properties The properties of the model. The properties are optional and can be null or
+   *     empty.
+   * @return The registered model object.
+   * @throws NoSuchSchemaException If the schema does not exist.
+   * @throws ModelAlreadyExistsException If the model already registered.
+   */
+  @Override
+  public Model registerModel(NameIdentifier ident, String comment, Map<String, String> properties)
+      throws NoSuchSchemaException, ModelAlreadyExistsException {
+    // TODO: preEvent
+    try {
+      Model model = dispatcher.registerModel(ident, comment, properties);
```

Comment:
This logic is hard to follow ...
In 'Model Event Dispatcher', we register a model?
And we have 'Model Dispatcher' as the parent class of this one?